backend/services/application.py

```diff
@@ -0,0 +1,230 @@
+from backend.entities.section_application_table import section_application_table
+from backend.entities.academics.section_entity import SectionEntity
+from backend.models.academics.section import Section
+from backend.models.application_details import (
+    New_UTADetails,
+    UserApplication,
+)
+from backend.models.user import User
+
+from backend.services.exceptions import ResourceNotFoundException
+
+from ..database import db_session
+
+__authors__ = ["Ben Goulet, Abdulaziz Al-Shayef"]
+__copyright__ = "Copyright 2024"
+__license__ = "MIT"
+
+
+class ApplicationService:
+    """ApplicationService is the access layer to TA applications."""
+
+    def __init__(self, session: Session = Depends(db_session)):
+        """Initializes a new ApplicationService.
+
+        Args:
+            session (Session): The database session to use, typically injected by FastAPI.
+        """
+        self._session = session
+
+    def list(self) -> list[New_UTADetails]:
+        """Returns all TA applications.
+
+        Returns:
+            list[New_UTA]: List of all current and previously submitted applications.
+        """
+        entities = self._session.query(ApplicationEntity).all()
+
+        return [entity.to_details_model() for entity in entities]
+
+        # implement list() for all types here later
+
+        # applications = []
+
+        # for entity in entities:
+        #     if entity.type == "new_uta": # type: ignore
+        #         applications.append(
+        #             entity.to_details_model()
+        #         )
+        #     elif entity.type == "returning_uta":
+        #         applications.append(entity.to_details_model())
+        #     elif entity.type == "uta":
+        #         applications.append(entity.to_details_model())
+        #     else:
+        #         applications.append(entity.to_model())
+
+        # return applications
+        
+    
+    def get_application(self, subject: User) -> UserApplication:
+        """Returns application(s) for a specific user
+
+        Returns:
+            list[Application]: List of all current applications for a specific user
+        """
+
+        application_entity = (
+            self._session.query(ApplicationEntity)
+            .filter(ApplicationEntity.user_id == subject.id)
+            .first()
+        )
+
+        if application_entity is None:
+            return None
+
+        sections = (
+            self._session.query(section_application_table)
+            .filter(section_application_table.c.application_id == application_entity.id)
+            .order_by(section_application_table.c.preference)
+            .all()
+        )
+
+        section_ids = [section[1] for section in sections]
+
+        sections_entity = []
+        for id in section_ids:
+            sections_entity.append(
+                self._session.query(SectionEntity)
+                .filter(SectionEntity.id == id)
+                .first()
+            )
```

Comment:
Small nit: Is it possible to do a more efficient search here? It should be possible to query based on a list of IDs, then remove duplicates (if needed) - this will help to limit the number of queries. Also, it might be better to break up this expression